daemon: Restore old CIDR identities

jrajahalme opened the pull request:
Dump old bpf IP cache on daemon startup and restore all locally allocated (CIDR) identities with the same numeric security identity as before. This helps avoid transient drops during restart due to identity changes.

Any restored identities that remain unused after 10 minutes are removed from the IP cache. This grace period is configurable with a new agent command line option `--identity-restore-grace-period` that takes a duration, e.g., `identity-restore-grace-period=5m` for 5 minutes.

Note that this feature requires support from the LInux kernel. Kernels from 4.11 to 4.15 inclusive do not support this feature.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
Locally allocated identities are now restored during restart, helping avoid transient drops due to identity changes in policies.
```

jrajahalme committed:
daemon: Restore old CIDR identities

Dump old ipcache on daemon startup and restore all locally allocated
(CIDR) identities with the same numeric security identity as before. This
helps avoid transient drops during restart due to identity changes.

Any restored identities that remain unused after 10 minutes are removed
from the IP cache. This grace period is configurable with a new agent
command line option --identity-restore-grace-period that takes a
duration, e.g., identity-restore-grace-period=5m for 5 minutes.

Note that this feature requires support from the Linux kernel. Kernels
from 4.11 to 4.15 inclusive do not support this feature.

Testing is done in the FQDN runtime tests to verify that locally
allocated identities are stable across restart, and that they also
reappear in selectorcache after FQDN policies are re-applied.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1768,6 +1771,16 @@ func runDaemon() {
+				// Release restored CIDR identities after a grace period (default 10
+				// minutes).  Any identities actually in use will still exist after
+				// this.
+				time.Sleep(option.Config.IdentityRestoreGracePeriod)
```
I don't understand this sleep. Why are we sleeping? Why can't we release the CIDR identities immediately after we have synchronized all state with kubernetes?

jrajahalme replied:
The node might be running without k8s, such as an external workload, etc.

jrajahalme replied:
Also, the identities for IPs via DNS proxy are only allocated when clients make the DNS requests. If those are wiped off the ipcache while the client is still in use there would be drops. Not sure if we are restoring DNS zombies on Cilium restart, though.

aanm replied:
What's the rational behind the 10 minutes? I think this might cause unexpected drops. Let's split the discussion in two ways:

> The node might be running without k8s, such as an external workload, etc.

I agree with this. We have `k8s.IsEnabled()` were we can detect if the node is running without k8s and add a sleep for external workloads. If we are running with k8s we could run this immediately afterwards all caches are synced.

> Also, the identities for IPs via DNS proxy are only allocated when clients make the DNS requests. If those are wiped off the ipcache while the client is still in use there would be drops. Not sure if we are restoring DNS zombies on Cilium restart, though.

and now regarding the IP via DNS proxy. IIRC those are stored in the endpoint, I believe one of these two fields contain that https://github.com/cilium/cilium/blob/6e7e9468be4b7a40e7c2b6adb8ba647aff1b6e33/pkg/endpoint/endpoint.go#L208-L214. After restoring all endpoints and k8s caches are synchronized couldn't we run `d.ipcache.ReleaseCIDRIdentitiesByCIDR(d.restoredCIDRs)`


Regardless what we decide here let's leave comments in the code stating why the time.Sleep is there.

joestringer replied:
I don't think that that DNSRules field on endpoint holds a reference to the identities, AFAIK they're held by the selectorcache. Though I have not traced the exact behaviour for DNS rules restoration. Based on the way that CIDR identity reference tracking happens from the proxy, I would expect that it should be enough to wait for k8s sync + full policy calculation (which endpoint regeneration should ensure).

aanm commented on pkg/identity/cache/local.go:
```diff
@@ -45,11 +45,22 @@ func (l *localIdentityCache) bumpNextNumericIdentity() {
-func (l *localIdentityCache) getNextFreeNumericIdentity() (identity.NumericIdentity, error) {
+func (l *localIdentityCache) getNextFreeNumericIdentity(idCandidate identity.NumericIdentity) (identity.NumericIdentity, error) {
+	// Try first with the given candidate
+	if idCandidate&identity.LocalIdentityFlag != 0 {
```
Nit, we could have a method for `NumericIdentity` which would return true or false if a certain identity was local. We already have a `IsReservedIdentity()` method.

jrajahalme replied:
Turns out the method already exists: `identity.HasLocalScope()`

jrajahalme replied:
Fixed.

aanm reviewed (changes requested):
There's a function named `IPCache.supportsDelete()` which is used to detect if delete is supported in the ipcache. We could create a similar function to detect if we can read from the ipcache. (The function could have the exact same behavior as `IPCache.supportsDelete()` but without the log messages that `IPCache.supportsDelete()` prints)

jrajahalme committed:
hubble: use HasLocalScope()

Use HasLocalScope() helper rather than open coding it.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
ipcache: Add SupportsDump() helper

Add SupportsDump() helper reusing the probing already done for SupportsDelete().

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aanm commented on pkg/maps/ipcache/ipcache.go:
```diff
@@ -210,13 +230,16 @@ func (m *Map) supportsDelete() bool {
+			m.deleteSupport = m.dumpSupport
 		}
 
 		if !m.deleteSupport {
```
if we run SupportsDelete and SupportsDump, won't the message on the line below be printed twice?

jrajahalme replied:
No, as the above code is executed via `sync.Once()`.

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1768,6 +1771,16 @@ func runDaemon() {
+				time.Sleep(option.Config.IdentityRestoreGracePeriod)
+				log.Debugf("Releasing reference counts for %d restored CIDR identities", len(d.restoredCIDRs))
+
+				d.ipcache.ReleaseCIDRIdentitiesByCIDR(d.restoredCIDRs)
```
I forgot on my initial review. `d.restoredCIDRs` is a slice which we don't know how big it is but I am assuming it can be big in certain clusters. After we are done with `d.ipcache.ReleaseCIDRIdentitiesByCIDR(d.restoredCIDRs)` can we can free up the memory usage by running `d.restoredCIDRs = nil`?

jrajahalme replied:
Done.

jrajahalme replied:
I made the sleep conditional on `!k8s.IsEnabled()` and adjusted the help text on the new CLI option accordingly.

jrajahalme replied:
Like this:
```
@@ -1776,10 +1776,20 @@ func runDaemon() {
 				// Release restored CIDR identities after a grace period (default 10
 				// minutes).  Any identities actually in use will still exist after
 				// this.
-				time.Sleep(option.Config.IdentityRestoreGracePeriod)
+				//
+				// This grace period is not needed when running with K8s, as in that
+				// case we get here only after all state has been sychronized from
+				// K8s. If running on an external workload, or in any other case
+				// where k8s is not configured we need a grace period during which
+				// policies can be restored by importing them usign the Cilium API.
+				if !k8s.IsEnabled() {
+					time.Sleep(option.Config.IdentityRestoreGracePeriod)
+				}
 				log.Debugf("Releasing reference counts for %d restored CIDR identities", len(d.restoredCIDRs))
 
 				d.ipcache.ReleaseCIDRIdentitiesByCIDR(d.restoredCIDRs)
+				// release the memory held by restored CIDRs
+				d.restoredCIDRs = nil
 			}
```

pchaigno commented on test/runtime/fqdn.go:
```diff
@@ -1087,7 +1119,30 @@ var _ = Describe("RuntimeFQDNPolicies", func() {
+		Expect(err).To(BeNil(), "ipcache can not be dumped")
+		equal, diff = checker.DeepEqual(ipcacheBefore, ipcacheAfterDNSPolicy)
+		Expect(equal).To(BeTrue(), "CIDR identities changed after policy import: %s", diff)
+		GinkgoPrint(fmt.Sprintf("Local scope identities in IP cache after re-import of DNS policy: %v", ipcacheAfterDNSPolicy))
```
Not asking for a change in this PR, but do we have FQDN coverage in the K8s tests and could we increase their coverage to cover this new case as well? AFAIK, we are still hoping to remove the Runtime tests sometime in the future.

jrajahalme replied:
Will address this in a follow-up.

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -466,9 +466,12 @@ func initializeFlags() {
+	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change when k8s is not enabled")
 	option.BindEnv(option.IdentityChangeGracePeriod)
 
+	flags.Duration(option.IdentityRestoreGracePeriod, defaults.IdentityRestoreGracePeriod, "Time to wait before releasing unused restored CIDR identities during agent restart")
```
Should we have `when k8s is not enabled` here as well?

jrajahalme replied:
I put that text in the wrong place, sorry. And now that is removed again as we likely need a grace period also with k8s (not 100% sure so be safe for now).

pchaigno commented on test/runtime/fqdn.go:
```diff
@@ -1013,6 +1014,27 @@ var _ = Describe("RuntimeFQDNPolicies", func() {
 		invalidIP := outsideIps[OutsideHttpd1]
 
+		// Apply a CIDR policy with some link local IPs first so that the DNS identities
+		// will not allocate from the beginning of the local identity range
```
And why do we want to avoid that?

jrajahalme replied:
So that we'd get different identities allocated after restart (where we do not import this CIDR policy) if the identity restoration did not work. Will add this to the comment.

pchaigno reviewed (approved):
LGTM!

It looks like there are many existing unit tests for the ipcache that could be extended to cover the new function arguments. It would probably be less of a pain to handle than a new end-to-end test, especially in Runtime :-/

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1768,6 +1771,26 @@ func runDaemon() {
+				// policies can be restored by importing them usign the Cilium API.
+				if !k8s.IsEnabled() {
+					time.Sleep(option.Config.IdentityRestoreGracePeriod)
+				}
```
What do we actually need to happen before we trigger the logic below? Full policy calculation for all endpoints?

Does "k8s is synchronized" also guarantee that Cilium has processed all of the events from k8s? Or only that Cilium has received all of the resources and has begun processing them? Given the way that policy is processed asynchronously in Cilium, I'm not sure that k8s sync will provide the guarantees we need around the reference counting of the CIDR identities that are referenced here :thinking:

jrajahalme replied:
@aanm Please resolve this with Joe. I made this sleep not to be done when k8s is enabled due to your change request.

jrajahalme replied:
Unless otherwise noted I'll remove the !k8s condition on the sleep to make sure we do not release these identities too soon.

jrajahalme replied:
Now sleeping unconditionally with this comment:
```
				// Release restored CIDR identities after a grace period (default 10
				// minutes).  Any identities actually in use will still exist after
				// this.
				//
				// This grace period is needed when running on an external workload
				// where policy synchronization is not done via k8s. Also in k8s
				// case it is prudent to allow concurrent endpoint regenerations to
				// (re-)allocate the restored identities before we release them.
```

jrajahalme replied:
We can tighten this down again later if necessary.

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -985,6 +1006,15 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 		realIdentityAllocator.InitIdentityAllocator(k8s.CiliumClient(), nil)
 
+		// Preallocate IDs for old CIDRs, must be called after InitIdentityAllocator
+		if option.Config.RestoreState && !option.Config.DryMode && len(d.restoredCIDRs) > 0 {
```
nit: Can we simplify this logic, since the other two conditions are implied by the population of `d.restoredCIDRs`?

```suggestion
		// Preallocate IDs for old CIDRs, must be called after InitIdentityAllocator
		if len(d.restoredCIDRs) > 0 {
```

jrajahalme replied:
Simplified :-)

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -466,9 +466,12 @@ func initializeFlags() {
 	option.BindEnv(option.FixedIdentityMapping)
 
-	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change")
+	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change when k8s is not enabled")
```
Is this meant to be changed or is this modifying the wrong parameter? This patch is about the `IdentityRestoreGracePeriod`, not the `IdentityChangeGracePeriod`.

jrajahalme replied:
changed the wrong parameter, will fix.

joestringer reviewed (changes requested):
Just one blocking nit since it looks like this PR changes the help text for an unrelated parameter.

I'm also not sure I understand how we guarantee that the release of the old identities is delayed long enough if k8s is enabled, see the top comment below.

joestringer reviewed (approved):
We should iterate a bit further on the sleep angle, but otherwise my previous feedback was addressed, thanks.
